llmfoundry/callbacks/dataset_swap_callback.py

```diff
@@ -104,11 +110,11 @@ def after_load(self, state: State, logger: Logger):
             self.all_dataset_configs.append(self.current_dataset_config)
 
     def state_dict(self):
-        return {
-            'dataset_index': self.dataset_index,
-            'all_dataset_configs': self.all_dataset_configs,
-        }
-
-    def load_state_dict(self, state: dict[str, Any]):
-        self.saved_dataset_index = state.get('dataset_index', 0)
-        self.all_dataset_configs = state.get('all_dataset_configs', [])
+        return DatasetSwapStateDict(
+            dataset_index=self.dataset_index,
+            all_dataset_configs=self.all_dataset_configs,
+        )
+
+    def load_state_dict(self, state: DatasetSwapStateDict):
+        self.saved_dataset_index = getattr(state, 'dataset_index', 0)
```

Comment:
@gupta-abhay this lgtm, but can you include manual test runs that succeeded for save & resume in the PR description? And add context to the PR description as well

Reply:
is there any example test I can take a look at for reference ?

Reply:
The runs you were trying before should work if you install foundry off this branch right?

Reply:
yes, I tested locally and it worked. Also doing a run through mcli and testing now. Give me like 10-15 mins to get back on that.

Reply:
ok my bad, I mistook what you were saying. I thought you meant to add a proper test within foundry for the callback testing vs just including runs with the fixes in place. Adding details now.